---
title: "5.1 Data Frames & 5.2 Data Files - Tutorial"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, error = TRUE, warning = TRUE)

library(ottr)
library(testthat)
```

---

At the end of this tutorial you should be able to:

* Create a data frame
* Index a data frame
* Modify a data frame
* Read CSV and TSV files
* Write a CSV file

---

### NOTE

This tutorial differs from previous tutorials in that section 5.1 & 5.2 will not have their separate tutorials. Instead parts of 5.2 are integrated throughout the sections in 5.1

---

## 5.1.2: Data Frame Creation

---

In this tutorial we will work with data from:
Enzel, P. et al. Comparative Metagenomics of Eight Geographically Remote Terrestrial Hot Springs. Microb Ecol. 2015 Aug. 70(2):411-24

This paper looks at the types of microorganisms living in 8 terrestrial hot springs. This table holds information about each of the hot springs examined.

| Sample     | Temperature (C) | pH  |
| ---------- | --------------- | --- |
| Ch2-EY65S  | 65	             | 7.0 |
| Sun Spring | 63	             | 5.9 |
| It6	       | 76	             | 3.0 |
| It3	       | 86	             | 5.5 |
| Is2-5S     | 88	             | 5.0 |
| Is3-13     | 90	             | 3.8 |
| CH1102     | 79	             | 1.8 |
| NL10       | 92	             | 3.5 |

* Note: Where a temperature range or pH range was given, the median temperature or pH is used.

Create a data frame called "hot_springs" with 3 columns named "sample", "temp", "pH", containing the information above. Use the appropriate data type for each column.
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W5.1.2.a.R")
```
**Check your answer by running the above block** 

Below are two vectors providing the location of each hot spring (in the same order as the hot springs are in the table above). One contains the city each hot spring is in (`cities`) and the other contains the country (`countries`). 

```{r}
cities <- c("Eryuan", "Uzon Caldera", "Pozzuoli", "Pisciarelli", "Grensdalur", "Krisuvik", "YNP", "YNP")
countries <- c("China", "Russia", "Italy", "Italy", "Iceland", "Iceland", "USA", "USA")
```

Use the `cbind()` function to add `cities` and `countries` to `hot_springs` (in that order). 
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W5.1.2.b.R")
```
**Check your answer by running the above block** 

Change the column names of `hot_springs` so that the last two columns are "city" and "country".
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W5.1.2.c.R")
```
**Check your answer by running the above block** 

Samples from each of these hot springs were sequenced and the sequencing reads were mapped to genomes to identify the proportion of reads that belong to different species. Here we will look at the total number of reads that were mapped (found to match a sequence in the genome of a sequenced species), and how many of those were mapped to a species in the Archea domain, and how many mapped to a species in the Bacteria domain.

A CSV file called "hot_springs_illumina_read_info.csv" contains information about sequencing reads for the four samples that were sequenced using Illumina sequencing technology.  

Use the `read.csv()` function to read in the data from the file and save it as a data frame called "illumina_read_info". 

Note: the file has a header and is located in the directory: "/home/jovyan/Week.5/5.2.Importing.Data.in.R/Tutorial.5.1.and.5.2".
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W5.1.2.d.R")
```
**Check your answer by running the above block** 

A TSV file called "hot_springs_454_read_info.txt" contains information about sequencing reads for the four samples that were sequenced using a different sequencing technology with much longer reads (called 454).  

Use the `read.table()` function to read in the data from the file and save it as a data frame called "s454_read_info". 

Note: the file has a header and is located in the directory: "home/jovyan/Week.5/5.2.Importing.Data.in.R/Tutorial.5.1.and.5.2".
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W5.1.2.e.R")
```
**Check your answer by running the above block** 

Combine `illumina_read_info` and `s454_read_info` into one data frame called "sequencing_reads".
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W5.1.2.f.R")
```
**Check your answer by running the above block** 

---

## 5.1.3: Data Frame Indexing

---

Calculate the mean hot spring temperature. Save the value to a variable named "mean_hot_spring_temp".
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W5.1.3.a.R")
```
**Check your answer by running the above block** 

Use any type of indexing to return a data frame containing the sample, city, and country names for rows 3 and 7 in `hot_springs`. Save the new data frame to a variable called "hot_spring_3_7_location".
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W5.1.3.b.R")
```
**Check your answer by running the above block** 

---

## 5.1.3: Data Frame Modification

---

Add a new column to `sequencing_reads` called "sequencing_tech". The first 4 values should be "illumina" and the last 4 values should be "454".
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W5.1.3.c.R")
```
**Check your answer by running the above block** 

Not all of the sequencing reads mapped to a Bacteria or Archea species, they may have mapped to a Eukaryotic or Viral species, or be unclassified. Subtract the columns `archea_reads` and `bacteria_reads` from `total_reads` to get the number of reads that mapped to something else to create a new vector called "other_reads".
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W5.1.3.d.R")
```
**Check your answer by running the above block** 

Add a new column called "other_reads" to `sequencing_reads` with the `other_reads` vector created above. 
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W5.1.3.e.R")
```
**Check your answer by running the above block** 

Write a new CSV file to the directory "home/jovyan/Week.5/5.2.Importing.Data.in.R/Tutorial.5.1.and.5.2" called "sequencing_reads.csv". The CSV file should contain the data frame `sequencing_reads`.
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W5.1.3.f.R")
```
**Check your answer by running the above block** 

To confirm the CSV was written correctly, read in the CSV you just created and save it to a variable called "my_CSV".
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W5.1.3.g.R")
```
**Check your answer by running the above block** 

---